### PR TITLE
Another attempt at improving performance

### PR DESCRIPTION
**Pull request by NeekSandhu**
Continuation of https://github.com/NeekSandhu/onigasm/pull/5

`benchmark.js` was run against `master`[includes #5] **(before)** and current patch **(after)**

This patch moves utf8<->utf16 offset mapping system from C++ to JS side, so that we could leverage automatic garbage collection in JS engines. This was neccessary as sometimes we need to re-use strings and computing encoding / utf offset mappings is very expensive.

Total time may look amazing in **After** image, but unfortunately, this patch has very weird and couter-intuitive behaviours (for example its terribly slow in cases where we'd expect it to be super-fast)

I'd love some help 😕

### Before

___

### CODE REVIEW

**Comment by NeekSandhu** [comment]
This patch continues as an overkill attempt at #9

File: benchmark.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        let stepFactor = charCount / 5
+        for (let i = 0; i < 100; i++) {
+            for (let j = 0; j < stepCount; j++) {
+                scanner.findNextMatchSync(str.slice(0, j * stepFactor))
+            }
+        }
+        log(`Uncached strings < ${charCount} characters\n:`, StopWatch.now() - t0, 'ms')
+    
+        t0 = StopWatch.now()
+        var onig_strs = Array(stepCount).map((_, i) => new OnigString(str.slice(0, stepFactor * i)))
```
Comment (aeschli) [inline benchmark.js]: `Array(stepCount).map(..)` doesn't work as expected (see here https://stackoverflow.com/questions/5501581/javascript-new-arrayn-and-array-prototype-map-weirdness_
Changing that to `Array(stepCount).fill(undefined).map(...` fixes the issue

**Comment by aeschli** [comment]
When fixing `Array(stepCount).map(..)` and increasing the number of different strings created to 50 (stepCount = 50), I get numbers that are more consistent:
```
Uncached strings < 100 characters
: 146 ms
Cached strings < 100 characters
: 13 ms

Uncached strings < 1000 characters
: 354 ms
Cached strings < 1000 characters
: 19 ms

Uncached strings < 10000 characters
: 3414 ms
Cached strings < 10000 characters
: 195 ms

Uncached strings < 50000 characters
: 26588 ms
Cached strings < 50000 characters
: 908 ms

Uncached strings < 100000 characters
: 56887 ms
Cached strings < 100000 characters
: 1919 ms

Uncached strings < 1000000 characters
: 566451 ms
Cached strings < 1000000 characters
: 20938 ms

Uncached strings < 2000000 characters
: 952051 ms
Cached strings < 2000000 characters
: 33723 ms
```

Pretty cool!

But if I'm not missing some important issue, the code is still too complex as I think the there's only one array for the mapping table needed, and that array can already be created in 'encode'.
The array index corresponds to the JavaScript string char offset.
- JS String: `Wörld`
- UTF-8 array: `\x57\xc3\xb6\x72\x6c\x64` 
- mappingTable = [0, 1, 3, 4, 5]
- stringToUTF8offset: offset => mappingTable[offset]
- UTF8ToStringOffset: utf8Offset => search for offset in array (ideally using binary search)

**Comment by NeekSandhu** [comment]
Well #9 merges encoding and offset table generation in just on `while(i < str.length) { }`.
In just one loop we're doing UTF8 encoding, UTF16, UTF16 > UTF8 map and UTF8 > UTF16 as well.

Unfortunately, after running some isolated benchmarks, it turns out that if all those "merged" tasks took `2500ms` for n runs, it took only `300ms` for just converting the string to UTF16 array.

I'm experimenting with encoding only UTF16, saving that on OnigString, and when ever `findBestMatch` is called first time it'll be treated differently. We pass UTF16 string to C runtime and have it encode to UTF8, plus generate offset maps, and return their respective pointers which we can copy from WASM heap to OnigString for subsequent calls.

But PR is welcome with your approach as I'm having hard time visualizing it _(using only one table w/ binary search that can be faster)_

**Comment by aeschli** [comment]
@NeekSandhu Here's my attempt: https://github.com/NeekSandhu/onigasm/pull/10 
It's slightly faster than my previous measurements, but I didn't had much time to measure and maybe there's potential to improve.

```
Uncached strings < 1000 characters
: 371 ms
Cached strings < 1000 characters
: 25 ms

Uncached strings < 10000 characters
: 3235 ms
Cached strings < 10000 characters
: 155 ms

Uncached strings < 50000 characters
: 19749 ms
Cached strings < 50000 characters
: 780 ms

Uncached strings < 100000 characters
: 39550 ms
Cached strings < 100000 characters
: 1531 ms
```

**Comment by NeekSandhu** [comment]
Please see https://github.com/NeekSandhu/onigasm/pull/10#issuecomment-398692902

**Pull request merged**